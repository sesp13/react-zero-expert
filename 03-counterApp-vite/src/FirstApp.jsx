import PropTypes from 'prop-types';

export const FirstApp = ({ title, subtitle, name }) => {
  return (
    <>
      <h1>{title}</h1>
      <p>{subtitle}</p>
      <p>{name}</p>
    </>
  );
};

FirstApp.propTypes = {
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.number.isRequired,
};

FirstApp.defaultProps = {
  name: 'Fernando herrera',
  subtitle: 'No hay subtítulo',
  title: 'No hay título',
};
