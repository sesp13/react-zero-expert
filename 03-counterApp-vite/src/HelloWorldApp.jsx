export function HelloWorldApp() {
  return <h1>Hello world!!!</h1>;
}
