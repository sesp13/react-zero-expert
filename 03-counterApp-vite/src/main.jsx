import React from 'react';
import ReactDOM from 'react-dom/client';
// import { HelloWorldApp } from './HelloWorldApp';
// import { FirstApp } from './FirstApp';
import { CounterApp } from './CounterApp';

// Import css
import './styles.css';

ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>
    {/* <HelloWorldApp /> */}
    {/* <FirstApp></FirstApp> */}
    <CounterApp value={0} />
  </React.StrictMode>
);
