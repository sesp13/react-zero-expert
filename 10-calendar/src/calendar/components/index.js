export * from './CalendarEvent';
export * from './CalendarModal';
export * from './FabAddNew';
export * from './Navbar';
