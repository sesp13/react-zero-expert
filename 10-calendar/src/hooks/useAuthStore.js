import { useSelector, useDispatch } from 'react-redux';
import { calendarApi } from '../api';
import { clearAuthErrorMessage, onChecking, onLogin, onLogout } from '../store';

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const startLogin = async ({ email, password }) => {
    try {
      dispatch(onChecking());
      const { data } = await calendarApi.post('/auth', { email, password });
      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      dispatch(onLogout('Invalid credentials'));
      setTimeout(() => {
        clearAuthErrorMessage();
      }, 10);
    }
  };

  const startRegister = async ({ name, email, password }) => {
    try {
      dispatch(onChecking());
      const { data } = await calendarApi.post('/auth/new', {
        name,
        email,
        password,
      });
      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      const errorMessage = error?.response?.data?.msg || 'Error on register';
      dispatch(onLogout(errorMessage));
      setTimeout(() => {
        clearAuthErrorMessage();
      }, 10);
    }
  };

  return {
    // properties
    status,
    user,
    errorMessage,
    // methods
    startLogin,
    startRegister,
  };
};
