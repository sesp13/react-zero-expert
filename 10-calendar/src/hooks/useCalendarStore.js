import { useSelector } from 'react-redux';

export const useCalendarStore = () => {
  const { events } = useSelector((store) => store.calendar);

  return {
    events,
  };
};
