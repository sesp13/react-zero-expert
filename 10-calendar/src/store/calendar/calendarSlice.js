import { createSlice } from '@reduxjs/toolkit';

export const calendarSlice = createSlice({
  name: 'calendar',
  initialState: {
    isLoadingEvents: true,
    events: [],
    activeEvent: null,
  },
  reducers: {
    onSelectActiveEvent: (state, { payload }) => {
      state.activeEvent = payload;
    },
    onAddNewEvent: (state, { payload }) => {
      state.events.push(payload);
      state.activeEvent = null;
    },
    onUpdateEvent: (state, { payload }) => {
      state.events = state.events.map((event) => {
        if (event._id === payload._id) {
          return payload;
        }
        return event;
      });
    },
    onDeleteEvent: (state) => {
      if (state.activeEvent) {
        state.events = state.events.filter(
          (event) => event._id !== state.activeEvent._id
        );
        state.activeEvent = null;
      }
    },
    onLoadEvents: (state, { payload = [] }) => {
      (state.isLoadingEvents = false),
        payload.forEach((event) => {
          const exists = state.events.some(
            (dbEvent) => dbEvent.id === event.id
          );
          if (!exists) {
            state.events.push(event);
          }
        });
    },
  },
});

export const {
  onSelectActiveEvent,
  onAddNewEvent,
  onUpdateEvent,
  onDeleteEvent,
  onLoadEvents,
} = calendarSlice.actions;
