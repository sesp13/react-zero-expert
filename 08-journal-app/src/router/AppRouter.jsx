import { Navigate, Route, Routes } from 'react-router-dom';
import { AuthRoutes } from '../auth/routes/AuthRoutes';
import { useCheckAuth } from '../hooks';
import { JournalRoutes } from '../journal/routes/JournalRoutes';
import { AuthStatus } from '../store/auth';
import { CheckingAuth } from '../ui';

export const AppRouter = () => {
  const { status } = useCheckAuth();

  if (status === 'checking') {
    return <CheckingAuth />;
  }

  return (
    <Routes>
      {status === AuthStatus.authenticated ? (
        <Route path="/*" element={<JournalRoutes />} />
      ) : (
        <Route path="auth/*" element={<AuthRoutes />} />
      )}
      <Route path="/*" element={<Navigate to="/auth/login" />} />
    </Routes>
  );
};
