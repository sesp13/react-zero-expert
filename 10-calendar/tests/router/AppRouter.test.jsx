import { render, screen } from '@testing-library/react';
import { useAuthStore } from '../../src/hooks/useAuthStore';
import { AppRouter } from '../../src/router/AppRouter';
import {
  authenticatedState,
  notAuthenticatedState,
} from '../fixtures/authStates';
import { MemoryRouter } from 'react-router-dom';
import { CalendarPage } from '../../src/calendar';

jest.mock('../../src/hooks/useAuthStore');
jest.mock('../../src/calendar', () => ({
  CalendarPage: () => <h1>CalendarPage</h1>,
}));

describe('Tests on <AppRouter />', () => {
  const mockCheckAuthToken = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should call the loading screen and call checkAuthToken', () => {
    useAuthStore.mockReturnValue({
      checkAuthToken: mockCheckAuthToken,
      status: 'checking',
    });
    render(<AppRouter />);
    expect(mockCheckAuthToken).toHaveBeenCalled();
    expect(screen.getByText('Loading')).toBeTruthy();
  });

  test('should show login if not authenticated', () => {
    useAuthStore.mockReturnValue({
      checkAuthToken: mockCheckAuthToken,
      status: notAuthenticatedState.status,
    });
    const { container } = render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );

    expect(screen.getByText('Ingreso'));
    expect(container).toMatchSnapshot();
  });

  test('should show the calendar if authenticated', () => {
    useAuthStore.mockReturnValue({
      checkAuthToken: mockCheckAuthToken,
      status: 'checking',
    });
    render(<AppRouter />);
    expect(mockCheckAuthToken).toHaveBeenCalled();
    expect(screen.getByText('Loading')).toBeTruthy();
  });

  test('should show login if not authenticated', () => {
    useAuthStore.mockReturnValue({
      checkAuthToken: mockCheckAuthToken,
      status: authenticatedState.status,
    });
    const { container } = render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );

    expect(screen.getByText('CalendarPage')).toBeTruthy();
  });
});
