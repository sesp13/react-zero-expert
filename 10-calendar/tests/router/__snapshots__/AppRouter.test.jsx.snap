// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <AppRouter /> should show login if not authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Repeat the password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Create account"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
